ENH: signal.ShortTimeFFT: determine arbitrary dual windows

DietBru opened the pull request:
The [ShortTimeFFT] can only calculate the canonical dual window (which is utilized in the [istft]). 

This PR provides a new function [closest_STFT_dual_window] to calculate the closest dual window in a least-squares sense for a given desired window as well as a new method [ShortTimeFFT.from_win_equals_dual] to determine windows which are equal to their dual window.  

The Constant OverLap Add (COLA) constraint is equivalent of having a constant dual window, i.e., ``all(ShortTimeFFT.dual_win == ShortTimeFFT.dual_win[0])``. Hence, `closest_STFT_dual_window()` is a generalization of [check_COLA], which is deprecated in this PR.

Closes #20608 by adding `'math:numref'` to the white list in [refguide_check.py].

Further minor changes:
* In  `_calc_dual_canonical_window(Fixed edgde case i)` fixed edge case for catching invalid zero-valued `win`. 
* Allow `np.int` for `hop` parameters.
* Removed minor typos and reference for STFT applications  
* Fixed a linter error in`_spectral_helper()`

Update: The rendered documentation can be found here:
* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [Updates]
* [check_COLA]

~~Update: Resolves  #20622 by adding additional documentation.~~
Update: Though there is an example added utilizing a cross-spectrogram, it is probably not expressive enough to close the Issue #20622.

[ShortTimeFFT]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.html

[istft]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.istft.html#scipy.signal.ShortTimeFFT.istft

[refguide_check.py]: https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L339

[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html

[check_COLA]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html#scipy.signal.check_COLA

[Updates]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

DietBru committed:
ENH: Determine arbitrary dual windows for short-time Fourier Transform

Provide a new function `closest_STFT_dual_window()` to calculate the closest dual window in a least-squares sense for a given desired window as well as a new method `ShortTimeFFT.from_win_equals_dual()` to determine windows which are equal to their dual window.

j-bowhay commented:
@sancholp would you be interested in reviewing this?

sancholp commented:
@j-bowhay yes, I'll take a look

sancholp commented on scipy/signal/_short_time_fft.py:
```diff
@@ -1130,6 +1133,12 @@ def spectrogram(self, x: np.ndarray, y: np.ndarray | None = None,
         ``None`` it needs to have the same shape as `x`.
 
+        The cross-spectrogram may be interpreted as the time-frequency analogon of the
+        cross-spectral density (consult `csd`). If the STFT is parametrized to be a
```
You could also briefly state the [coherency spectrum](https://en.wikipedia.org/wiki/Cross-spectrum#:~:text=Squared%20coherency%20spectrum%5Bedit%5D), another useful spectrogram that can be obtained using this functionality.

sancholp reviewed (approved):
The documentation changes are fine. Maybe somewhere you could explain that the csd spectrogram can also be used to create a [coherence spectrogram](https://en.wikipedia.org/wiki/Cross-spectrum#:~:text=Squared%20coherency%20spectrum%5Bedit%5D).

sancholp commented:
@larsoner @ilayn @rgommers would it be possible to get this merged?

DietBru commented:
Thanks for looking this, @sancholp. I added a few words explaining a coherence spectrogram in the last commit.

On an unrelated side note: The Wikipedia [definition] of the squared coherency spectrum you cited cannot be extended into a time-frequency version of a spectrogram.  The reason being that the definition is based on wide-sense stationary stochastic processes for which time-frequency analysis (as opposed to spectral analysis) is not useful.
It makes more sense to interpret it as a normalized time-frequency [cross-correlation] function.



[definition]: https://en.wikipedia.org/wiki/Cross-spectrum#Squared_coherency_spectrum
[cross-correlation]: https://en.wikipedia.org/wiki/Cross-correlation

DietBru commented:
The test failure seems to be unrelated.

Update: The rendered documentation can be found here:
* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [Updates] in Tutorial section
* [check_COLA]

[ShortTimeFFT]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.html

[istft]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.istft.html#scipy.signal.ShortTimeFFT.istft

[refguide_check.py]: https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L339

[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html#scipy.signal.ShortTimeFFT.from_win_equals_dual

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html#scipy.signal.ShortTimeFFT.spectrogram

[check_COLA]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html#scipy.signal.check_COLA

[Updates]: https://output.circle-artifacts.com/output/job/0d22762e-0f9b-4e9b-b09e-0a062172470f/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

j-bowhay commented:
@DietBru would you be able to resolve the merge conflict

DietBru commented:
Thanks for the pointer, @j-bowhay. The current rendered documentation can be found here:
* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [Updates] in Tutorial section
* [check_COLA]

[ShortTimeFFT]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.html

[istft]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.istft.html
[refguide_check.py]: https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L339

[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html
[check_COLA]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html

[Updates]: https://output.circle-artifacts.com/output/job/e9588664-39a0-4b2a-a042-a84c603c0e6d/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

DietBru committed:
Miscellaneous documentation improvements

DietBru committed:
Fixed linter warnings

DietBru commented:
Rebased to current main and updated documentation a little bit—the rendered docs can be found here:

* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [Updates] in Tutorial section
* [check_COLA]



[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/a4c60b41-7281-42bf-949d-f568a3f104dc/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/a4c60b41-7281-42bf-949d-f568a3f104dc/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/a4c60b41-7281-42bf-949d-f568a3f104dc/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html

[check_COLA]: https://output.circle-artifacts.com/output/job/a4c60b41-7281-42bf-949d-f568a3f104dc/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html

[Updates]: https://output.circle-artifacts.com/output/job/a4c60b41-7281-42bf-949d-f568a3f104dc/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

j-bowhay commented:
@DietBru sorry I you already have and I've just forgotten but do you mind posting about this on the dev forum since this adds a new function to the public api

DietBru committed:
Replace `assert_allclose` with `xp_assert_close` in `test_short_time_fft.py`

DietBru commented:
Good point, @j-bowhay—I have forgotten to post as well. The forum post can be found [here](https://discuss.scientific-python.org/t/additional-functionality-for-scipy-signal-shorttimefft-determine-arbitrary-dual-windows/1489)

DietBru committed:
Added unitary scaling to ShortTimeFFT and various documentation improvements mainly about unitary mappings.

Also:
* Some fixed double backticks in docstrs
* Insured that L2-norm typesetting corresponds to Wikipedia's typesetting convention.
* Further small documentation improvements

DietBru commented:
Rebased to current main and improved the handling of unitary mappings—the rendered docs can be found here:

* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [Updates] in Tutorial section
* [check_COLA]



[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html

[check_COLA]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html

[Updates]: https://output.circle-artifacts.com/output/job/2fbf1eab-d6ef-45a5-97b8-0163de6e680b/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

larsoner commented:
Yes but I waited too long now the CircleCI renders don't show up :scream: 

I'll merge `main` into this branch but @DietBru feel free to force-push over it with a rebase if you prefer

larsoner committed:
Merge remote-tracking branch 'upstream/main' into STFT_EnhancedDualWindows

* upstream/main: (605 commits)
  DOC: Fix documentation for ``predecessors`` matrix in ``shortest_path``, ``bellman_ford``, ``johnson`` and ``dijkstra`` (#22388)
  STY: fix lint rule UP031 (#21987)
  MAINT: special: A bit of clean up in stirling2.h (#22381)
  DOC: sparse.linalg: fix doctest in scipy.sparse.linalg._norm.py (#22375)
  DOC: Replace link to X in header with link to scientific python forum (#22380)
  DOC: integrate: fix quad documentation to correctly describe QAGS algorithm (#22366)
  BUG: special: Fix unchecked malloc in stirling2.h (#22339)
  DOC: interpolate.AAA: add may vary to example
  DOC: sparse.linalg: sparray updates in doc_strings and Sakurai and MikotaK methods (#22376)
  BUG: sparse.linalg.norm: fix return type (#22372)
  TST: bump tolerance on TestHyp2f1.test_region3[hyp2f1_test_case23] (#22364)
  BUG: signal.bilinear handles complex input, and strips leading zeros (#22313)
  Small code formatting improvement in scipy/signal/_fir_filter_design.py
  MAINT: remove temporary `# type: ignore`'s from #22162
  MAINT: Stop installing libhighs
  MAINT: signal/windows: f64 degenerate windows, for consistency
  TST: interpolate: small tolerance bump to TestAAA.test_basic_functions
  TST: signal: update to dpss updates
  TST: signal/windows: fix tests
  TST: signal/windows: add explicit dtypes to asarrays
  ...

larsoner commented on scipy/signal/_spectral_py.py:
```diff
@@ -997,7 +997,24 @@ def spectrogram(x, fs=1.0, window=('tukey', .25), nperseg=None, noverlap=None,
 def check_COLA(window, nperseg, noverlap, tol=1e-10):
-    r"""Check whether the Constant OverLap Add (COLA) constraint is met.
+    r"""Check whether the Constant OverLap Add (COLA) constraint is met
+    (leagcy function).
```
```suggestion
    (legacy function).
```

larsoner commented on doc/source/tutorial/signal.rst:
```diff
@@ -1756,26 +1760,148 @@ zeros, enlarging :math:`\vb{U}` so all slices which touch :math:`x[k]` contain
-    w_d[m] = w[m] \inv{\sum_{\eta\in\IZ} \big|w[m + \eta\, h]\big|^2}\ .
+    w_d[m] = w[m] \inv{\sum_{\eta\in\IZ} \big|w[m + \eta\, h]\big|^2}
+           =\ w[m] \inv{\sum_{l=0}^{M-1} \big|w[l]\big|^2 \delta_{l+m,\eta h}}\ ,
+             \quad \eta\in \IZ\ .
```
Split over two (aligned-on-`=`) lines? It overflows for me at least

![image](https://github.com/user-attachments/assets/c451322d-4c63-42a2-98bf-cf6721bc3320)

DietBru replied:
Done. I wish there would be scroll bars for formulas like for the code blocks. The problem is that in the current configuration the text width is allowed to vary to some amount (can be seen by changing the width of the browser window).

larsoner commented on scipy/signal/_short_time_fft.py:
```diff
@@ -73,6 +73,153 @@ class method ``ShortTimeFFT.from_dual()``.
     return win / DD
 
 
+def closest_STFT_dual_window(win: np.ndarray, hop: int, desired_dual: np.ndarray, *,
```
So it looks like the vast majority of uses in examples and tests are of the form:
```
closest_STFT_dual_window(win, hop, np.ones(len(win))
```
with a few select others others being
```
closest_STFT_dual_window(np.ones(len(desired_dual)), hop, desired_dual)
```
i.e., one of `win` or (much more frequently) `desired_dual` is `np.ones(len(the_other))`.

Is it worth making this explicit by changing the signature to:
```
desired_dual : np.ndarray | None = None
```
? So given a `win` and `hop` you can more easily get a dual that is closest to a `rect` (which seems like the by-far most common case)?

DietBru replied:
I agree with you: Rectangular windows a probably the most common use case. Done in commit 0142714d7f.

larsoner commented on scipy/signal/_short_time_fft.py:
```diff
@@ -430,15 +600,159 @@ def from_window(cls, win_param: str | tuple | float,
         """Window function as real- or complex-valued 1d array.
 
-        This attribute is read only, since `dual_win` depends on it.
+        This attribute is read-only, since `dual_win` depends on it.
```
Just had a thought... yes in principle it's read-only but in practice someone could be bitten by:
```
x = stft.win
x[:] = 0  # or whatever, could happen in downstream code where a user passes `x` and it gets mutated
assert (stft.win == 0).all()
```
We could either set the array flags on `self._win` or (easier) return `self._win.copy()` here.

Not related to this PR so feel free to ignore but might be worthwhile while you're in there anyway...

DietBru replied:
Thanks for the suggestion—if haven't thought of that. 
Let me think about that a bit. 
I would prefer to not add this here, since PR is long enough already :smiley:.

larsoner commented on scipy/signal/_spectral_py.py:
```diff
@@ -997,7 +997,24 @@ def spectrogram(x, fs=1.0, window=('tukey', .25), nperseg=None, noverlap=None,
+        >>> w, w_rect, hop = windows.hann(12, sym=False), np.ones(12), 6
+        >>> dual_win, alpha = closest_STFT_dual_window(w, hop, w_rect, scaled=True)
+        >>> np.allclose(dual_win/alpha, w_rect, atol=1e-10, rtol=0)
+        True
```
For a "this is equivalent" example it would be nice to have the `check_COLA` call here that returns `True` as well
```suggestion
        True
        >>> check_COLA(window, len(w), len(w) - hop)
        True
```
or whatever the correct invokation is :slightly_smiling_face:

larsoner commented on scipy/signal/tests/test_short_time_fft.py:
```diff
@@ -59,6 +59,123 @@ def test__calc_dual_canonical_window_exceptions():
+    d_u, u = closest_STFT_dual_window(win * scale_fac, nperseg - noverlap,
+                                      desired_dual, scaled=False)
+
+    assert u == 1, "Scaling factor not 1 for parameter `scaled=True`!"
```
```suggestion
    # this should be hard-coded not computed, so no need for allclose
    assert u == 1, "Scaling factor not 1 for parameter `scaled=True`!"
```

larsoner commented on scipy/signal/tests/test_short_time_fft.py:
```diff
@@ -222,6 +339,78 @@ def test_from_window(win_params, Nx: int):
+        # noinspection PyTypeChecker
+        ShortTimeFFT.from_win_equals_dual(w, hop=2, fs=1, scale_to='invalid')
+
+@pytest.mark.parametrize('fft_bins', (True, False))
```
Obligatory pedantic two-newlines-between-func-defs :slightly_smiling_face: 
```suggestion

@pytest.mark.parametrize('fft_bins', (True, False))
```

larsoner reviewed (commented):
Okay, I think I might have understood it well enough to give some useful comments @DietBru. Looks good to me otherwise!

DietBru committed:
In file `_short_time_fft.py`:
* Make the default of parameter `desired_dual` in function `ShortTimeFFT.closest_STFT_dual_window` a rectangular window.
* Improve docstr of `ShortTimeFFT.from_win_equals_dual`.
* Add "Parameters" and "Returns" section to the docstrs of `ShortTimeFFT.stft_detrend` and `ShortTimeFFT.spectrogram`.
* Add types to "Parameters" section in `ShortTimeFFT.stft`, `ShortTimeFFT.stft_detrend` and `ShortTimeFFT.spectrogram`.
* Make the example in the `ShortTimeFFT.spectrogram` docstr more copy and paste friendly.

In file `test_short_time_fft`:
* Added test for testing detrending in `ShortTimeFFT.stft_detrend`.
* Add clarifying comment to `test_closest_STFT_dual_window_cola` test.

Furthermore:
* Improve legacy message in the docstr of `signal.check_COLA`.
* Shorten width of Eq. `eq_STFT_CanonDualWin` in file `signal.rst`.

DietBru commented on scipy/signal/_short_time_fft.py:
```diff
@@ -430,15 +600,159 @@ def from_window(cls, win_param: str | tuple | float,
+        `from_win_equals_dual()`, then the resulting window would be
+        ``SFT.win * np.sqrt(SFT.mfft) = 6``, i.e., the same window which is
+        plotted below.
+        """
```
Good point. Reworked the docstr in commit  0142714d7f a little bit to make it clearer.

DietBru commented on scipy/signal/_short_time_fft.py:
```diff
@@ -430,15 +600,159 @@ def from_window(cls, win_param: str | tuple | float,
+        mapping, i.e., dividing the window by ``np.sqrt(mfft)`` and multiplying the
+        dual window by the same amount.
+
+        All other parameters have the identical meaning as in the initializer.
```
Done in commit 0142714d7f.

DietBru commented on scipy/signal/_short_time_fft.py:
```diff
@@ -861,14 +1183,24 @@ def spectrogram(self, x: np.ndarray, y: np.ndarray | None = None,
         `stft_detrend`, hence all parameters are discussed there. If `y` is not
         ``None`` it needs to have the same shape as `x`.
 
+        The cross-spectrogram may be interpreted as the time-frequency analogon of the
```
You are right. It's also missing in `stft_detrend. 

Added those in commit 0142714d7f.

DietBru committed:
Increased tolerance in `test_compare_stft_detrend` due to CI failing.

DietBru commented:
Thank @larsoner for taking it upon you to do the review and for the feedback. I hope I sufficiently addressed all your points.

The failing test seems unrelated. The updated rendered documentation can be found here:

* [closest_STFT_dual_window]
* [ShortTimeFFT.from_win_equals_dual]
* [ShortTimeFFT.spectrogram]
* [ShortTimeFFT.stft_detrend]
* [ShortTimeFFT.stft]
* [Updates] in Tutorial section
* [check_COLA]



[closest_STFT_dual_window]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.closest_STFT_dual_window.html

[ShortTimeFFT.from_win_equals_dual]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.from_win_equals_dual.html

[ShortTimeFFT.spectrogram]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html

[ShortTimeFFT.stft]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.stft.html

[ShortTimeFFT.stft_detrend]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.ShortTimeFFT.stft_detrend.html


[check_COLA]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/reference/generated/scipy.signal.check_COLA.html

[Updates]: https://output.circle-artifacts.com/output/job/0f7b8d99-96f8-4c63-9678-4af828c43b1a/artifacts/0/html/tutorial/signal.html#equation-eq-stft-canondualwin

larsoner commented:
Awesome, thanks @DietBru !
